l2-contracts/contracts/bridge/L2SharedBridge.sol

```diff
@@ -39,169 +40,172 @@ contract L2SharedBridge is IL2SharedBridge, Initializable {
-        ) {
-            revert InvalidCaller(msg.sender);
-        }
-
-        address expectedL2Token = l2TokenAddress(_l1Token);
-        address currentL1Token = l1TokenAddress[expectedL2Token];
-        if (currentL1Token == address(0)) {
-            address deployedToken = _deployL2Token(_l1Token, _data);
-            if (deployedToken != expectedL2Token) {
-                revert AddressMismatch(expectedL2Token, deployedToken);
-            }
-
-            l1TokenAddress[expectedL2Token] = _l1Token;
+    function finalizeDeposit(bytes32 _assetId, bytes memory _data) public override onlyL1Bridge {
+        address assetHandler = assetHandlerAddress[_assetId];
+        if (assetHandler != address(0)) {
+            IL2AssetHandler(assetHandler).bridgeMint(L1_CHAIN_ID, _assetId, _data);
         } else {
-            if (currentL1Token != _l1Token) {
-                revert AddressMismatch(_l1Token, currentL1Token);
-            }
+            IL2AssetHandler(nativeTokenVault).bridgeMint(L1_CHAIN_ID, _assetId, _data);
+            assetHandlerAddress[_assetId] = address(nativeTokenVault);
         }
 
-        IL2StandardToken(expectedL2Token).bridgeMint(_l2Receiver, _amount);
-        emit FinalizeDeposit(_l1Sender, _l2Receiver, expectedL2Token, _amount);
-    }
-
-    /// @dev Deploy and initialize the L2 token for the L1 counterpart
-    function _deployL2Token(address _l1Token, bytes calldata _data) internal returns (address) {
-        bytes32 salt = _getCreate2Salt(_l1Token);
-
-        BeaconProxy l2Token = _deployBeaconProxy(salt);
-        L2StandardERC20(address(l2Token)).bridgeInitialize(_l1Token, _data);
-
-        return address(l2Token);
+        emit FinalizeDepositSharedBridge(L1_CHAIN_ID, _assetId, keccak256(_data));
     }
 
     /// @notice Initiates a withdrawal by burning funds on the contract and sending the message to L1
     /// where tokens would be unlocked
-    /// @param _l1Receiver The account address that should receive funds on L1
-    /// @param _l2Token The L2 token address which is withdrawn
-    /// @param _amount The total amount of tokens to be withdrawn
-    function withdraw(address _l1Receiver, address _l2Token, uint256 _amount) external override {
-        if (_amount == 0) {
-            revert AmountMustBeGreaterThanZero();
-        }
-
-        IL2StandardToken(_l2Token).bridgeBurn(msg.sender, _amount);
-
-        address l1Token = l1TokenAddress[_l2Token];
-        if (l1Token == address(0)) {
-            revert EmptyAddress();
+    /// @param _assetId The L2 token address which is withdrawn
+    /// @param _assetData The data that is passed to the asset handler contract
+    function withdraw(bytes32 _assetId, bytes memory _assetData) public override {
+        address assetHandler = assetHandlerAddress[_assetId];
+        bytes memory _bridgeMintData;
+
+        /// should the address not be set already? Or do we need this to do it automatically.
```

Comment:
A good question : ) 

I think that similarly to withdrawals on L1, we should not handle special case of native toens

Reply:
It should be, removed the conditions, lets see what breaks